engine/apps/alerts/tasks/check_escalation_finished.py

```diff
@@ -1,48 +1,130 @@
+        )
+
+    task_logger.info(f"{base_msg} passed the audit checks")
+
+
+def get_auditable_alert_groups_started_at_range() -> typing.Tuple[datetime.datetime, datetime.datetime]:
     """
-    This task periodically checks if there are no alert groups with not finished escalations.
-    TODO: QA this properly, check if new type of escalations had been added
+    NOTE: this started_at__range is a bit of a hack..
+    we wanted to avoid performing a migration on the alerts_alertgroup table to update
+    alert groups where raw_escalation_snapshot was None. raw_escalation_snapshot being None is a legitimate case,
+    where the alert group's integration does not have an escalation chain associated with it.
+
+    However, we wanted a way to be able to differentiate between "actually None" and "there was an error writing to
+    raw_escalation_snapshot" (as this is performed async by a celery task).
+
+    This field was updated, in the commit that added this comment, to no longer be set to None by default.
+    As part of this celery task we do a check that this field is in fact not None, so if we were to check older
+    alert groups, whose integration did not have an escalation chain at the time the alert group was created
+    we would raise errors
     """
+    return (datetime.datetime(2023, 2, 5), timezone.now() - timezone.timedelta(days=2))
+
+
+# don't retry this task as the AlertGroup DB query is rather expensive
+@shared_task
+def check_escalation_finished_task():
     AlertGroup = apps.get_model("alerts", "AlertGroup")
     AlertReceiveChannel = apps.get_model("alerts", "AlertReceiveChannel")
 
-    CHECKING_TOLERANCE = timezone.timedelta(minutes=5)
-    CHECKING_TIME = timezone.now() - CHECKING_TOLERANCE
-
-    alert_groups = AlertGroup.all_objects.filter(
+    alert_groups = AlertGroup.all_objects.using(get_random_readonly_database_key_if_present_otherwise_default()).filter(
         ~Q(channel__integration=AlertReceiveChannel.INTEGRATION_MAINTENANCE),
         ~Q(silenced=True, silenced_until__isnull=True),  # filter silenced forever alert_groups
         is_escalation_finished=False,
         resolved=False,
         acknowledged=False,
         root_alert_group=None,
-        estimate_escalation_finish_time__lte=CHECKING_TIME,
+        started_at__range=get_auditable_alert_groups_started_at_range(),
     )
 
     if not alert_groups.exists():
-        return
-
-    exception_template = "Escalation for alert_group {} is not finished at expected time {}, now {}"
-
-    now = timezone.now()
-    exception_text = "\n".join(
-        exception_template.format(alert_group.pk, alert_group.estimate_escalation_finish_time, now)
-        for alert_group in alert_groups
-    )
+        task_logger.info("There are no alert groups to audit, everything is good :)")
 
-    ids = alert_groups.values_list("pk", flat=True)
-    task_logger.debug(ids)
+    for alert_group in alert_groups:
+        audit_alert_group_escalation(alert_group)
```

Comment:
As far as I understand this checking will stop after finding the first "stuck" alert group. So we won't collect the ids "stuck" of alert groups, right?

Reply:
good point. you're correct, currently this would raise on the first stuck alert group, and skip checking the rest. I'll modify this

Reply:
this should be handled in [this commit](https://github.com/grafana/oncall/pull/1266/commits/ef7fce9ba0c7e5e5a42eb77b0cc552033b2c8871). Added a test for this scenario.